rfcs/widget-restructure-rfc-1.md

```diff
@@ -0,0 +1,385 @@
+    
+    // Now we can make this properly notify the context
+    this.add_class("foo");
+  }
+}
+```
+
+### TL;DR
+
+1. Generate `MyWidgetProps`
+2. Move rendering to associated function
+3. Create `MyWidgetContext` to interface between render function and `MyWidget` and `KayakContext`
+4. ???
+5. Profit
+
+## Alternatives
+
+### Faith
+
+The biggest alternative, again, is just to trust that the developer isn't going to cause any issues and that we make it clear what code *could* cause issues. But I think we can still do something to help. And coming up with a solution also improves the overall experience and API.
+
+### Data-less Widgets
+
+One big alternative to a portion of this is to just not store data on the widget itself. All relevant data for a widget will be stored in the `KayakContext`. This forces all code that needs a bit of data about a widget to also need access to the current context.
+
+This could make code outside the functional component much more cumbersome to use.
+
+Additionally, we run into the issue of storing the widget's ID, since that data does need to exist on the widget regardless.
+
+### Context Setters
+
+One of the issues I briefly mentioned was about the idea that updating a widget's own data does not directly alert the context. This could be solved by just having users write the notification code:
+
+```rust
+#[widget]
+fn Foo(context: &mut KayakContext) {
+  self.add_class("bar");
+  context.notify(KayakNotify::ClassAdded); // or however it's implemented
+}
+```
+
+From an API perspective, this isn't great. Using the Widget Context idea, we can move that required code into the interface method:
+
+```rust
+impl<'a> FooContext<'a> {
+  pub fn add_class(&mut self, class: &str) {
+    self.widget.add_class(class);
+    context.notify(KayakNotify::ClassAdded);
+  }
+}
+
+#[widget]
+fn Foo() {
+  this.add_class("bar");
+}
+```
+
+Also note that we can't do this:
+
+```rust
+#[widget]
+fn Foo(context: &mut KayakContext) {
+  context.add_class("bar");
+}
+```
+
+Since this requires that `Foo` have a public method `add_class` for the context to set. And the alternatives for those issues go back to the ones listed above.
+
+### Move Render Function into Widget Context
+
+One potential alternative to what's mentioned above would be to move the associated render function from `MyWidget` into the `MyWidgetContext`. This would allow `self` to be used as normal and even give direct access to `context` and `widget` if needed, while still granting the separation that allows us to add custom logic.
+
+## Sample Generated Output
+
+<details>
+<summary><em>Show/Hide Code</em></summary>
+
+```rust
+use crate::OnEvent;
+use crate::render_command::RenderCommand;
+use crate::styles::{Style, StyleProp};
+
+mod my_widget_module {
+    use derivative::Derivative;
+    use crate::{Color, Index, KayakContext, OnEvent, Widget};
+    use crate::styles::Style;
+
+    #[derive(Derivative)]
+    #[derivative(Default, Debug, PartialEq)]
```

Comment:
Do we still need `PartialEq` here? 🤔

Reply:
I'm not sure. I basically copied it from a macro expansion on the `TextBox` widget.